src/overlays/actors/ovl_En_Rat/z_en_rat.c

```diff
@@ -28,123 +47,913 @@ const ActorInit En_Rat_InitVars = {
+ */
+void EnRat_ChooseDirection(EnRat* this) {
+    Vec3f sp74;
+    s16 angle;
+
+    if (this->actionFunc != EnRat_Idle) {
+        angle = this->actor.yawTowardsPlayer - this->actor.shape.rot.y;
+        if (this->axisUp.y < -0.25f) {
+            angle -= 0x8000;
+        }
+    } else {
+        if (Actor_DistanceToPoint(&this->actor, &this->actor.home.pos) > 50.0f) {
+            Vec3f sp64;
+            Vec3f sp58;
+            Vec3f sp4C;
+            Vec3f sp40;
+
+            Matrix_RotateZS(-this->actor.home.rot.z, MTXMODE_NEW);
+            Matrix_RotateXS(-this->actor.home.rot.x, MTXMODE_APPLY);
+            Matrix_RotateYS(-this->actor.home.rot.y, MTXMODE_APPLY);
+            Matrix_MultVec3f(&this->axisUp, &sp40);
+            Math_Vec3f_Sum(&this->actor.world.pos, &this->axisForwards, &sp64);
+            Matrix_MultVec3f(&sp64, &sp4C);
+            Matrix_MultVec3f(&this->actor.home.pos, &sp64);
+            Matrix_MultVec3f(&this->actor.world.pos, &sp58);
+            angle = Math_Vec3f_Yaw(&sp58, &sp64) - Math_Vec3f_Yaw(&sp58, &sp4C);
+            if (sp40.y < -0.25f) {
+                angle -= 0x8000;
+            }
+
+            angle += (s16)randPlusMinusPoint5Scaled(2048.0f);
+        } else {
+            angle = (Rand_ZeroOne() < 0.1f) ? (s16)randPlusMinusPoint5Scaled(2048.0f) : 0;
+        }
+    }
+
+    angle = CLAMP(angle, -0x800, 0x800);
+    Matrix_RotateAxisF(angle * 0.0000958738f, &this->axisUp, MTXMODE_NEW);
+    Matrix_MultVec3f(&this->axisForwards, &sp74);
+    Math_Vec3f_Copy(&this->axisForwards, &sp74);
+    Math3D_CrossProduct(&this->axisUp, &this->axisForwards, &this->axisLeft);
+    this->shouldRotateOntoSurfaces = true;
+}
+
+/**
+ * Returns true if the Real Bombchu is on a collision poly or is on the water's surface.
+ */
+s32 EnRat_IsOnCollisionPoly(PlayState* play, Vec3f* posA, Vec3f* posB, Vec3f* posResult, CollisionPoly** poly,
+                            s32* bgId) {
+    WaterBox* waterBox;
+    s32 isOnWater;
+    f32 waterSurface;
+
+    if ((WaterBox_GetSurface1(play, &play->colCtx, posB->x, posB->z, &waterSurface, &waterBox)) &&
+        (waterSurface <= posA->y) && (posB->y <= waterSurface)) {
+        isOnWater = true;
+    } else {
+        isOnWater = false;
+    }
+
+    if (BgCheck_EntityLineTest1(&play->colCtx, posA, posB, posResult, poly, 1, 1, 1, 1, bgId)) {
+        if (!(func_800C9A4C(&play->colCtx, *poly, *bgId) & 0x30) && (!isOnWater || (waterSurface <= posResult->y))) {
+            return true;
+        }
+    }
+
+    if (isOnWater) {
+        posResult->x = posB->x;
+        posResult->y = waterSurface;
+        posResult->z = posB->z;
+        *poly = NULL;
+        *bgId = BGCHECK_SCENE;
+        return true;
+    }
+
+    return false;
+}
+
+/**
+ * "Floor" in this case means floors, walls, and ceilings, since the Real Bombchu can run on any of them.
+ */
+s32 EnRat_IsTouchingFloor(EnRat* this, PlayState* play) {
```

Comment:
Isn't all surfaces floors, walls, ceilings? So could this be `EnRat_IsTouchingSurface`? Or if there's surfaces that aren't included, could you emphasize those in the description?